### PR TITLE
[WIP] Added restarter

### PR DESCRIPTION
**Pull request by geekodour**
See #154 

Updating a metadata in the manifest and applying update restarts the prometheus instances.

### CODE REVIEW

**Comment by krasi-georgiev** [comment]
Thanks
haven't looked at the code yet, but just to be clear it needs to restart both instances at the same time and should kill the prometheus binary and not a graceful restart

**Comment by krasi-georgiev** [comment]
about the DCO 
https://github.com/prometheus/prometheus/wiki/DCO-signing

**Comment by geekodour** [comment]
I think k8s sends a `SIGTERM` so might be a graceful restart. I am not sure, I'll look into it and make changes.

**Comment by krasi-georgiev** [comment]
yes it needs to be a kill signal and it should be at the same time on both instance and that both should be started again at the same time so that both graph align properly after the restart.

**Comment by krasi-georgiev** [comment]
Is there a blocker here?

**Comment by geekodour** [comment]
yes #184 .

If #184 gets merged the same restart script will work for both the instances as they are run in the same way. this PR is not yet review ready I think.

**Comment by geekodour** [comment]
Still have to add restart on the tsdb metrics, otherwise this is working fine. 

Alternative way to do the same will use `terminationGracePeriodSeconds:0` and then applying a new deployment. Let me know if that'll be more preferable than this.

also, should I squash the commits to one?

File: components/prombench/manifests/benchmark/3b_prometheus-test.yaml [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+data:
+  restarter.sh: |
+    #!/bin/sh
+    echo "Restarting prometheus"
+    kill -SIGHUP $(pidof start.sh)
+  killer.sh: |
+    #!/bin/sh
+    echo "Killing prometheus"
+    kill -9 $(pidof prometheus)
+  fetchtini.sh: |
```
Comment (geekodour) [inline components/prombench/manifests/benchmark/3b_prometheus-test.yaml]: we can move the tini download part to the prometheus-builder image
then run prometheus-builder for both deployments

- the release deployment should only copy tini to the required directory
- the pr deployment will work same + copy tini to the required directory

this way we won't need to run this script from k8s and reduce the number of
initContainers by 1 for pr deployment

File: components/prombench/manifests/benchmark/3_prometheus-test.yaml [hunk 2]
```diff
@@ -698,6 +698,68 @@ data:
...
+    echo "Downloading tini init system"
+
+    if [ -z $VOLUME_DIR ]; then
+        echo "environment variables not set correctly, exiting."
+        exit 1;
+    fi
+
+    wget -O $VOLUME_DIR/tini \
+    https://github.com/krallin/tini/releases/download/v0.18.0/tini-static
+    chmod +x $VOLUME_DIR/tini
```
Comment (geekodour) [inline components/prombench/manifests/benchmark/3_prometheus-test.yaml:727]: we can move the tini download part to the prometheus-builder image
then run prometheus-builder for both deployments

- the release deployment should only copy tini to the required directory
- the pr deployment will work same + copy tini to the required directory

this way we won't need to run this script from k8s and reduce the number of
initContainers by 1 for pr deployment

**Comment by krasi-georgiev** [comment]
> Still have to add restart on the tsdb metrics, otherwise this is working fine.

Remind me why do we need these metrics?

> Alternative way to do the same will use terminationGracePeriodSeconds:0 and then applying a new deployment. Let me know if that'll be more preferable than this.

Do you mean to do something like - delete deployment without waiting for a grace shutdown and recreate the deployment?
if this send a `kill -9` signal than yes It looks like the easier approach.

> also, should I squash the commits to one?

no need I will do it when merging.

**Comment by geekodour** [comment]
though this approach gives much more granularity using the approach in #187 is simpler once `preStop` hook is functional. can this be closed then?

**Comment by krasi-georgiev** [comment]
can you sum up the main difference, as I haven't looked at the code yet.

**Comment by geekodour** [comment]
#180 (this) uses WebSockets to exec commands inside the containers. In this case, the restarter program has complete control over when the `SIGKILL` is sent and when Prometheus is restarted. 

#187 uses the fact that Deployments will restore the pod to its desired state. So the restarter program only updates the Deployment spec and uses  `strategy.type: Recreate` and [`preStop`](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#container-hooks) hook to send `SIGKILL` to existing Pod when the deployment is applied and once the container is killed. Kubernetes brings it back up so we don't have to worry about restarting the prometheus here. 

None of #180 or #187 change `terminationGracePeriodSeconds` so it remains the same in the manifest files.

Both of these use [tini init](https://github.com/krallin/tini/) so that we can run Prometheus as non PID1 process so that we can send the `SIGKILL`

**Comment by geekodour** [comment]
Closing for the same reason as #187

**Pull request closed without merging**